File: Year-2023/LeetCode-February-Challenge/MinimizeDeviationinArray.java
import java.util.*;
class Solution {
    public int minimumDeviation(int[] nums) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        int minValue = Integer.MAX_VALUE;

        for(int x: nums){
            if(x % 2 == 0){
                pq.add(x);
                minValue = Math.min(minValue, x);
            } else {
                pq.add(x * 2);
                minValue = Math.min(minValue, x * 2);
            }
        }

        int res = Integer.MAX_VALUE;
        while(!pq.isEmpty()){
            int top = pq.poll();
            res = Math.min(res, top - minValue);
            if(top % 2 != 0) break;
            minValue = Math.min(minValue, top / 2);
            pq.add(top / 2);
        }
        return res;
    }
}
public class MinimizeDeviationinArray{
    public static void main(String[] args) {
        int[] nums = {1,2,3,4};
        Solution s = new Solution();
        System.out.println(s.minimumDeviation(nums));
    }
}

File: Year-2023/LeetCode-February-Challenge/MinimizeDeviationinArray.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        priority_queue<int> pq;
        int minValue = INT_MAX;
        for(auto x: nums){
            if(x % 2) pq.push(x * 2), minValue = min(minValue, x * 2);
            else pq.push(x), minValue = min(minValue, x);
        }
        int res = INT_MAX;
        while(!pq.empty()){
            int top = pq.top();
            pq.pop();
            res = min(res, top - minValue);
            if(top % 2 != 0) break;
            minValue = min(minValue, top / 2);
            pq.push(top / 2);
        }
        return res;
    }
};
int main(){
    vector<int> nums = {1,2,3,4};
    Solution *s = new Solution();
    cout<<s->minimumDeviation(nums)<<endl;
    return 0;
}